### PR TITLE
NMS-13806: Improve requisition interface IP validation 

### PR DESCRIPTION
**Pull request by amay94**
### All Contributors

* [x] Have you read our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [x] Have you (electronically) signed the [OpenNMS Contributor Agreement](https://cla-assistant.io/OpenNMS/opennms)?

### Contribution Checklist

* Please [make an issue in the OpenNMS issue tracker](https://issues.opennms.org) if there isn't one already.<br />Once there is an issue, please:
  1. update the title of this PR to be in the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
  4. once you've created this PR, please link to it in a comment in the JIRA issue
  Don't worry if this sounds like a lot, we can help you get things set up properly.
* If this is a new or updated feature, is there documentation for the new behavior?
* If this is new code, are there unit and/or integration tests?
* If this PR targets a `foundation-*` branch, does it try to avoid changing files in `$OPENNMS_HOME/etc/`?

### What's Next?

A PR should be assigned at least 2 reviewers.  If you know that someone would be a good person to review your code, feel free to add them.

If you need help making additions or changes to the documentation related to your changes, please let us know.

In any case, if anything is unclear or you want help getting your PR ready for merge, please don't hesitate to say something in the comments here,
or in [the #opennms-development chat channel](https://chat.opennms.com/opennms/channels/opennms-development).

Once reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the OGP) you are welcome to merge the PR when you're ready.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13806
* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-13805

### CODE REVIEW

File: opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java [hunk 1]
```diff
@@ -318,9 +330,10 @@ public String getIpAddr() {
      */
     public void setIpAddr(String value) {
         try {
-            m_ipAddress = InetAddressUtils.toIpAddrString(InetAddressUtils.getInetAddress(value));
+            m_ipAddress = InetAddressUtils.getInetAddress(value);
+            m_ipAddressStr = value;
         } catch (Throwable e) {
-            throw new IllegalArgumentException("Invalid IP address specified", e);
+            throw new IllegalArgumentException(String.format("Invalid IP address specified: {}", value), e);
```
Comment (christianpape) [inline opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java:336]: `{}` is not a String.format() specifier.

File: opennms-provision/opennms-provision-persistence/src/test/java/org/opennms/netmgt/provision/persist/FilesystemForeignSourceRepositoryTest.java [hunk 2]
```diff
@@ -87,6 +87,13 @@ public void testRequisition() throws Exception {
         assertEquals("node name matches", "apknd", v.getNodeReqs().get(0).getNodeLabel());
         
         m_foreignSourceRepository.delete(r);
+
+        // This is a queueing repo, so the delete task is submitted to an Executor. Wait it out.
+        try {
+            Thread.currentThread().sleep(500);
```
Comment (christianpape) [inline opennms-provision/opennms-provision-persistence/src/test/java/org/opennms/netmgt/provision/persist/FilesystemForeignSourceRepositoryTest.java:93]: Use await().atMost().until()...

**Comment by amay94** [comment]
Still need to add:
* Send an event when a JSON RequisitionInterface is invalid
* Log the invalid IP address clearly at a level higher than DEBUG (per NMS-13806, bullet 3)

**Review by RangerRick — APPROVED** [review]

File: opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java [hunk 3]
```diff
@@ -498,4 +517,35 @@ public int compareTo(final RequisitionInterface other) {
...
+     *                               does not resolve to a valid address
+     */
+    protected void validateHost() throws IPValidationException {
+        try {
+            m_ipAddress = new IPAddress(m_ipAddressStr).toInetAddress();
+        }
+        catch (IllegalArgumentException iae) {
+            final Event e = EventUtils.createInterfaceRejectedEvent("Provisiond", m_ipAddressStr);
+            LOG.warn("Invalid IP address found in a requisition interface: {}", m_ipAddressStr);
+            StaticEventForwarder.getStaticEventForwarder().sendNow(e);
```
Comment (cgorantla) [inline opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java:547]: I don't like the way event forwarder is wired here.  We shouldn't wire in any dependencies in model classes.  Clients that load `RequisitionInterface` should handle validation and try to send event.

Comment (amay94) [reply opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java:547]: I tried to catch the validation issues in other locations, but there are code paths to ingest Requisition objects from the filesystem that I couldn't get to reliably validate the object before using it. Specifically, [this test](https://github.com/OpenNMS/opennms/blob/73e9ad89621078003d629444d7c90a46600f2378/opennms-provision/opennms-provision-persistence/src/test/java/org/opennms/netmgt/provision/persist/FilesystemForeignSourceRepositoryTest.java#L112) was continually failing when comparing the local Requistion to the one gotten from the repository.

The desired behavior from [NMS-13806](https://issues.opennms.org/browse/NMS-13806) is to just ignore any invalid interfaces and remove them from the model before ingesting. Throwing an error back to a client trying to ingest a large amount of nodes would violate this, as the parsing would be stopped. I tried to do all of the validation in the validate() methods that all of the Requisition model objects have, but the test I reference above was failing, and I couldn't figure out why the Requisition retrieved from the repository wasn't getting properly validated. 

The only reliable way to catch the issue that i found was to have the [ValidationEventHandler](https://github.com/OpenNMS/opennms/pull/4195/files#diff-6929f4c77301a59258c4200dd4197c145eebe41c4f5e61c15d3cddd923664afcL98) catch it and attempt to continue parsing, so the issue is only known within the model object.

I don't necessarily like the wiring either, but this is the only way I could think of to get an event out with this information in it. Another option is to drop the event entirely and just log the issue and move on.

Comment (cgorantla) [reply opennms-provision/opennms-provision-persistence/src/main/java/org/opennms/netmgt/provision/persist/requisition/RequisitionInterface.java:547]: I would opt to remove sending event from model objects. Catch specific validation exception and log it with `WARN` and move on.

File: opennms-webapp-rest/src/main/java/org/opennms/web/rest/v1/RequisitionRestService.java [hunk 4]
```diff
@@ -451,7 +452,7 @@ public Response addOrReplaceRequisition(@Context final UriInfo uriInfo, final Re
             LOG.debug("error validating incoming requisition with foreign source '{}'", requisition.getForeignSource(), e);
             throw getException(Status.BAD_REQUEST, e.getMessage());
         }
-        debug("addOrReplaceRequisition: Adding requisition %s (containing %d nodes)", requisition.getForeignSource(), requisition.getNodeCount());
+        debug("addOrReplaceRequisition: Adding requisition {} (containing {} nodes)", requisition.getForeignSource(), requisition.getNodeCount());
```
Comment (cgorantla) [inline opennms-webapp-rest/src/main/java/org/opennms/web/rest/v1/RequisitionRestService.java:455]: Just replace them with `LOG.debug`

**Review by cgorantla — COMMENTED** [review]
Thanks for the update @amay94.

Looks good to me.

It should be possible to directly test this validation on `RequisitionInterface` with `RequisitionFrom18Test`
Add an invalid ip address  in  test resources and number of imported interfaces should still be same.

**Comment by cgorantla** [comment]
> Thanks for the update @amay94.
> 
> Looks good to me.
> 
> It should be possible to directly test this validation on `RequisitionInterface` with `RequisitionFrom18Test` Add an invalid ip address in test resources and number of imported interfaces should still be same.

While you are there, update these two lines.
```
        m_foreignSourceRepository.setForeignSourcePath(this.getClass().getResource("/empty").getPath());
        m_foreignSourceRepository.setRequisitionPath(this.getClass().getResource("/1.8-upgrade-test").getPath());
```

**Review by christianpape — APPROVED** [review]
Looks good to me.

File: opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/operations/SaveOrUpdateOperation.java [hunk 5]
```diff
@@ -128,23 +123,23 @@ public void foundInterface(String ipAddr, Object descr, final PrimaryType primar
         //interface to the node.
         m_node.addIpInterface(m_currentInterface);
     }
-	
+
```
Comment (fooker) [inline opennms-provision/opennms-provisiond/src/main/java/org/opennms/netmgt/provision/service/operations/SaveOrUpdateOperation.java:126]: Same here

**Review by fooker — CHANGES_REQUESTED** [review]

File: core/api/src/main/java/org/opennms/core/network/IPValidationException.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+package org.opennms.core.network;
+
+import javax.xml.bind.ValidationException;
+
+/**
+ *  Sublcass of the generic XML ValidationException that indicates
+ *  that an IP address parsed from an XML document is invalid.
+ *
+ *  See <code>RequisitionNode.validate()</code>
```
[the remark this answers is not in the document]
Comment (amay94) [reply core/api/src/main/java/org/opennms/core/network/IPValidationException.java]: updated

File: core/api/src/main/java/org/opennms/core/network/InetAddressXmlAdapter.java [hunk 7]
```diff
@@ -28,12 +28,17 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.net.InetAddress;
 
 import javax.xml.bind.annotation.adapters.XmlAdapter;
 
 public class InetAddressXmlAdapter extends XmlAdapter<String, InetAddress> {
 
+    private static Logger LOG = LoggerFactory.getLogger(InetAddressXmlAdapter.class);
```
[the remark this answers is not in the document]
Comment (amay94) [reply core/api/src/main/java/org/opennms/core/network/InetAddressXmlAdapter.java]: done

**Review by fooker — DISMISSED** [review]

File: core/api/src/main/java/org/opennms/core/network/InetAddressXmlAdapter.java [hunk 8]
```diff
@@ -43,7 +48,13 @@ public String marshal(final InetAddress inetAddr) throws Exception {
     /** {@inheritDoc} */
     @Override
     public InetAddress unmarshal(final String ipAddr) throws Exception {
-        return (ipAddr == null || ipAddr.isEmpty())? null : new IPAddress(ipAddr).toInetAddress();
+        try {
+            return (ipAddr == null || ipAddr.isEmpty()) ? null : new IPAddress(ipAddr).toInetAddress();
+        }
+        catch (Throwable t) {
+            LOG.warn(String.format("Invalid IP address <%s>", ipAddr));
```
Comment (cgorantla) [inline core/api/src/main/java/org/opennms/core/network/InetAddressXmlAdapter.java:55]:
```suggestion
           LOG.warn("Invalid IP Address {}", ipAddr);
```

File: core/xml/src/main/java/org/opennms/core/xml/JaxbUtils.java [hunk 9]
```diff
@@ -71,6 +71,7 @@
 
 import org.apache.commons.io.IOUtils;
 import org.eclipse.persistence.jaxb.MarshallerProperties;
+import org.opennms.core.network.IPValidationException;
```
Comment (cgorantla) [inline core/xml/src/main/java/org/opennms/core/xml/JaxbUtils.java:74]: Can be removed

**Review by cgorantla — COMMENTED** [review]
Minor comments otherwise looks good.

So overall does it now work with both xml and json requisitions ?
When there is invalid IPAddress,  does the whole requisition skipped or only that specific IPAddress ?

**Comment by amay94** [comment]
@cgorantla Yes, invalid interfaces are ignored for both JSON and XML input, as long as there is a RequisitionNode object containing the interface. If the request contains just a single invalid interface, the request fails.

**Review by cgorantla — APPROVED** [review]

**Comment by amay94** [comment]
@cgorantla no changes since the last review, just rebasing on top of the latest develop

**Review by cgorantla — APPROVED** [review]

**Pull request merged**